CHAINCODE_AS_A_SERVICE_TUTORIAL.md

```diff
@@ -0,0 +1,197 @@
+
+- The docker image for the peer contains a builder for chaincode-as-a-service preconfigured. This is named 'ccaasbuilder'. This removes the need to build your own external builder and repackage and configure the peer
+- The `ccaasbuilder` applications are included in the binary tgz archive download for use in other circumstances. The `sampleconfig/core.yaml` is updated as well to refer to 'ccaasbuilder'
+- The 2.4.1 Java Chaincode release has been updated to remove the need to write a custom bootstrap main class, similar  to the NodeJS Chaincode. It is intended that this will be added to the go chaincode as well.
+
+## End-to-end with the the test-network
+
+The `test-network` and some of the chaincodes have been updated to support running chaincode-as-a-service. The commands below assume that you've got the latest fabric-samples cloned, along with the latest Fabric docker images. 
+
+It's useful to have two terminal windows open, one for starting the Fabric Network, and a second for monitoring all the docker containers.
+
+In your 'monitoring' window, run this to watch all activity from the all the docker containers on the `fabric_test` network
+
+```bash
+# from the fabric-samples repo
+./test-network/monitordocker.sh
+```
+In the 'Fabric Network' window, start the test network
+
+```bash
+cd test-network
+./network.sh up createChannel
+```
+
+You can run other variants of this command, eg to use CouchDB or CAs, without affecting the '-as-a-service' feature. The three keys steps are:
+
+- Build a docker image of the contract. Both `/asset-transfer-basic/chaincode-typescript` and `/asset-transfer-basic/chaincode-java` have been updated with Dockerfiles
+- Install, Approve and Commit a chaincode definition. This is unchanged, but the chaincode package contains connection information (hostname,port,tls certificates etc.), not code
+- Start the docker container(s) containing the contract
+
+Note that the order listed isn't mandatory. The key thing is that the containers are running before the first transaction is set by the peer. Remember that this could be on the `commit` if the `initRequired` flag is set.
+
+This sequence can be run as follows
+```bash
+./network.sh deployCCAAS  -ccn basicts -ccp ../asset-transfer-basic/chaincode-typescript 
+```
+
+This is very similar to the `deployCC` command, it needs the name, and path. But also needs to have the port the chaincode container is going use. As each container is on the `fabric-test` network, you might wish to alter this so there are no collisions with other chaincode containers.
+
+You should be able to see the contract starting in the monitoring window. There will be two containers running, one for org1 and one for org2. The container names contain the organzation/peer and the name of the chaincode.
+
+To test things are working you can invoke the 'Contract Metadata' function.
+
+```bash
+# set the environment variables to work as org1
+export $(./setOrgEnv.sh org1 | xargs)
```

Comment:
Ok, the chaincode gets built and deployed now when running latest PR.
However, this command gives an error:
`./setOrgEnv.sh: line 24: ${ORG,,}: bad substitution`
and then I'm unable to do the query.
But if I setup my command line per the regular instructions:
https://hyperledger-fabric.readthedocs.io/en/latest/test_network.html#interacting-with-the-network
it works!

Reply:
The ${,,} syntax is I think Bash v4

Reply:
@mbwhite I don't think macs come with Bash v4, would be best to avoid friction here.

I'd suggest it is anyways better to just copy/paste the environment variables from https://hyperledger-fabric.readthedocs.io/en/latest/test_network.html#interacting-with-the-network , this way users actually learn what is required rather than accepting the script magic.

Reply:
Why MacOs comes with a 10 year old version of Bash is a mystery; I've swapped to the environment variables as per that link